src/libraries/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.Linux.cs

```diff
@@ -147,656 +104,784 @@ private void FinalizeDispose()
-
-            // For the Created and Deleted events, we need to always
-            // register for the created/deleted inotify events, regardless
-            // of the supplied filters values. We explicitly don't include IN_DELETE_SELF.
-            // The Windows implementation doesn't include notifications for the root directory,
-            // and having this for subdirectories results in duplicate notifications, one from
-            // the parent and one from self.
-            result |=
-                Interop.Sys.NotifyEvents.IN_CREATE |
-                Interop.Sys.NotifyEvents.IN_DELETE;
-
-            // For the Changed event, which inotify events we subscribe to
-            // are based on the NotifyFilters supplied.
-            const NotifyFilters filtersForAccess =
-                NotifyFilters.LastAccess;
-            const NotifyFilters filtersForModify =
-                NotifyFilters.LastAccess |
-                NotifyFilters.LastWrite |
-                NotifyFilters.Security |
-                NotifyFilters.Size;
-            const NotifyFilters filtersForAttrib =
-                NotifyFilters.Attributes |
-                NotifyFilters.CreationTime |
-                NotifyFilters.LastAccess |
-                NotifyFilters.LastWrite |
-                NotifyFilters.Security |
-                NotifyFilters.Size;
-            if ((filters & filtersForAccess) != 0)
-            {
-                result |= Interop.Sys.NotifyEvents.IN_ACCESS;
-            }
-            if ((filters & filtersForModify) != 0)
-            {
-                result |= Interop.Sys.NotifyEvents.IN_MODIFY;
-            }
-            if ((filters & filtersForAttrib) != 0)
-            {
-                result |= Interop.Sys.NotifyEvents.IN_ATTRIB;
-            }
+            // Guards the watchers of the inotify instance.
+            public static readonly object s_watchersLock = new();
 
-            // For the Rename event, we'll register for the corresponding move inotify events if the
-            // caller's NotifyFilters asks for notifications related to names.
-            const NotifyFilters filtersForMoved =
-                NotifyFilters.FileName |
-                NotifyFilters.DirectoryName;
-            if ((filters & filtersForMoved) != 0)
+            private static INotify? s_currentInotify;
+
+            public static Watcher? StartWatcher(FileSystemWatcher fsw)
             {
-                result |=
-                    Interop.Sys.NotifyEvents.IN_MOVED_FROM |
-                    Interop.Sys.NotifyEvents.IN_MOVED_TO;
-            }
+                Watcher watcher;
+                lock (s_watchersLock)
+                {
+                    // If there is no running instance, start one.
+                    if (s_currentInotify is null || s_currentInotify.IsStopped)
+                    {
+                        INotify inotify = new(s_watchersLock);
+                        inotify.Start();
```

Comment:
Right here we are in a situation where we have created new `INotify` with the `INotifyInit`, and we call `INotify.Start` which creates a new background thread that starts processing the events in a loop:

https://github.com/dotnet/runtime/blob/3cb75b023f945216f52ad70b76b6786b6473aa3d/src/libraries/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.Linux.cs#L208

Are we not going to hit the following condition?

https://github.com/dotnet/runtime/blob/3cb75b023f945216f52ad70b76b6786b6473aa3d/src/libraries/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.Linux.cs#L504-L507

I would expect the call to `Start` after a call to `s_currentInotify.CreateWatcherCore(fsw);`

Reply:
The `watchersLock` is held in `StartWatcher` until we've tried to add a watch.

Then when the thread start method runs, it obtains the lock (therefore: waiting for `StartWatcher`) to check if a watch was effectively added. This causes the thread to stop in the unlikely case where we failed to add a watch for the FSW's base path.